packages/typespec-client-generator-core/src/decorators.ts

```diff
@@ -165,18 +165,26 @@ export const $client: ClientDecorator = (
   }
   const explicitName = options?.properties.get("name")?.type;
   const name: string = explicitName?.kind === "String" ? explicitName.value : target.name;
-  const explicitService = options?.properties.get("service")?.type;
-  const service =
-    explicitService?.kind === "Namespace"
-      ? explicitService
-      : (findClientService(context.program, target) ?? (target as any));
+  let service = options?.properties.get("service")?.type;
 
-  if (!isService(context.program, service)) {
+  if (service?.kind !== "Namespace") {
+    service = findClientService(context.program, target);
+  }
+
+  if (
+    service === undefined ||
+    service.kind !== "Namespace" ||
+    !isService(context.program, service) ||
+    !service.decorators.some(
+      (d) => d.definition?.name === "@service" && d.definition?.namespace.name === "TypeSpec",
```

Comment:
Curious on what happened to `isService` from compiler lib.

nit, if we need write this test twice (also in `findClientService`), we may use a util function.

Reply:
`isService` depend on the decorator apply sequence. it may return empty.